### PR TITLE
new-log-viewer: Implement log-level filtering.

### PR DESCRIPTION
**Pull request by davemarco**
# References
new-log-viewer series: #45 #46 #48 #51 #52 #53 #54 #55 #56 #59 #60 #61 #62

# Description
* Implemented buildIdx() to do deserialization in JSONL decoder. Previously, deserialization was done inside constructor. In addition, now we are parsing log level inside in buildIdx() and not in decode(). 
* Implement log level filtering with JSON logs.
* TODO: CLP IRv1 log-level filtering is to be added once the equivalent is implemented in clp-ffi-js. The implementation in clp-ffi-js will start once we agree with the decoder handling in this PR.

# Validation performed
1. Referred to #46 , launched debug server.
2. Referred to #54, tested page switching with Monaco editor with filters off.
3. Referred to #54, tested page switching with single (TRACE) log level filter.
4. Referred to #54, tested page switching with multiple (TRACE, DEBUG) log level filter.

### CODE REVIEW

File: new-log-viewer/src/services/LogFileManager.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import {
     Decoder,
     DecoderOptionsType,
+    FULL_RANGE_END_IDX
```
Comment (junhaoliao) [inline new-log-viewer/src/services/LogFileManager.ts:4]: lint:
```suggestion
    FULL_RANGE_END_IDX,
```

File: new-log-viewer/src/services/LogFileManager.ts [hunk 2]
```diff
@@ -175,6 +176,11 @@ class LogFileManager {
             } due to a limitation in Chromium-based browsers.`);
         }
 
+        const result = decoder.buildIdx(0, FULL_RANGE_END_IDX)
+        if (result != null && result.numInvalidEvents > 0) {
```
Comment (junhaoliao) [inline new-log-viewer/src/services/LogFileManager.ts:180]: lint:
```suggestion
        if (null !== result && 0 < result.numInvalidEvents) {
```
1. Let's use "yodo" conditions, where we put constants / literals on the left of the comparison operators.
2. In JavaScript, the difference between `!=` and `!==` is that the former does "loose inequality" comparison, where type conversions happen before the values are compared. The later one compares the type as well. Here's a good reference: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness

Comment (junhaoliao) [inline new-log-viewer/src/services/LogFileManager.ts:180]: Maybe the correct logic is to check against `null` equality and use `||`
```suggestion
        if (null === result || 0 < result.numInvalidEvents) {
```

File: new-log-viewer/src/services/LogFileManager.ts [hunk 3]
```diff
@@ -175,6 +176,11 @@ class LogFileManager {
             } due to a limitation in Chromium-based browsers.`);
         }
 
+        const result = decoder.buildIdx(0, FULL_RANGE_END_IDX)
```
Comment (junhaoliao) [inline new-log-viewer/src/services/LogFileManager.ts:179]: `buildIdxResult`?

File: new-log-viewer/src/services/LogFileManager.ts [hunk 4]
```diff
@@ -175,6 +176,11 @@ class LogFileManager {
...
+        if (result != null && result.numInvalidEvents > 0) {
+            console.error(`${result.numInvalidEvents} were invalid.`);
```
Comment (junhaoliao) [inline new-log-viewer/src/services/LogFileManager.ts:181]: Would it be better to print the result out (as an object) directly so that we can better understand how many results are valid & how many are invalid?

File: new-log-viewer/src/services/decoders/JsonlDecoder.ts [hunk 5]
```diff
@@ -4,6 +4,7 @@ import {
     DecodeResultType,
     JsonlDecoderOptionsType,
     LogEventCount,
+    FULL_RANGE_END_IDX,
```
Comment (junhaoliao) [inline new-log-viewer/src/services/decoders/JsonlDecoder.ts:7]: lint:
```suggestion
    DecodeResultType,
    FULL_RANGE_END_IDX,
    JsonlDecoderOptionsType,
    LogEventCount,
```

File: new-log-viewer/src/services/decoders/JsonlDecoder.ts [hunk 6]
```diff
@@ -16,6 +17,14 @@ import {
...
 import LogbackFormatter from "../formatters/LogbackFormatter";
 
+/**
+ * Parsed JSONL.
+ * TODO: Add timestamp to interface by parsing in buildIdx and remove functionality from decode.
+ */
+interface JsonLogEvent {
+    level: LOG_LEVEL
+    jsonLog: JsonObject
+}
```
Comment (junhaoliao) [inline new-log-viewer/src/services/decoders/JsonlDecoder.ts:27]: Would it be better to add a LogEvent interface for all types of decoders, and here we extend the interface to include JSON logs specific properties? i.e. `level`s and `timestamp`s are generally expected for all types of logs, right?

File: new-log-viewer/src/services/decoders/JsonlDecoder.ts [hunk 7]
```diff
@@ -45,24 +56,59 @@ class JsonlDecoder implements Decoder {
...
     }
 
     buildIdx (beginIdx: number, endIdx: number): Nullable<LogEventCount> {
-        // This method is a dummy implementation since the actual deserialization is done in the
-        // constructor.
 
-        if (0 > beginIdx || this.#logEvents.length < endIdx) {
-            return null;
+        if (beginIdx !== 0 || endIdx !== FULL_RANGE_END_IDX) {
+            throw new Error("Partial range deserialization is not yet supported.");
```
Comment (junhaoliao) [inline new-log-viewer/src/services/decoders/JsonlDecoder.ts:69]: Shall we free the `this.#dataArray` at the end of this method?

File: new-log-viewer/src/services/decoders/JsonlDecoder.ts [hunk 8]
```diff
@@ -16,6 +17,14 @@ import {
 } from "../../typings/logs";
 import LogbackFormatter from "../formatters/LogbackFormatter";
```
Comment (junhaoliao) [inline new-log-viewer/src/services/decoders/JsonlDecoder.ts:19]: lint:
```suggestion


```

**Review by junhaoliao — CHANGES_REQUESTED** [review]
Great job! I took a brief look and the code looks mostly good to me. Let's clean up the code with ESLint and we can do a more thorough round of review.

File: new-log-viewer/src/services/decoders/JsonlDecoder.ts (hunk 6 above)
Comment (junhaoliao) [reply new-log-viewer/src/services/decoders/JsonlDecoder.ts:28]: > The logEvents are private to the implementation so they don't necessarily need to be exposed to the interface.

That's a fair point. Let me trace the code and see what makes more sense.

File: new-log-viewer/package.json [hunk 9]
```diff
@@ -29,21 +29,26 @@
     "@babel/preset-env": "^7.24.7",
     "@babel/preset-react": "^7.24.7",
     "@babel/preset-typescript": "^7.24.7",
+    "@eslint/js": "^9.9.1",
```
Comment (junhaoliao) [inline new-log-viewer/package.json:32]: Is there a reason for adding these dependencies? Supposedly the equivalents have been listed as peer dependencies of `eslint-config-yscope` and those are automatically installed if you're using npm@7 or higher.

As an aside, most IDEs has built-in support (or via plugins) to auto lint files on save. Here's a sample config for enabling ESLint auto lint in PyCharm: 
image

File: new-log-viewer/src/components/Layout.tsx [hunk 10]
```diff
@@ -165,15 +173,27 @@ const handleLogEventNumInputChange = (ev: React.ChangeEvent<HTMLInputElement>) =
...
+        numFilteredEvents,
         pageNum,
+
+        changeLogLevelFilter,
+        loadFile,
     } = useContext(StateContext);
     const {logEventNum} = useContext(UrlContext);
 
+    const [selectedLogLevels, setSelectedLogLevels] =
+        useState<number[]>(LOG_LEVEL_NAMES_LIST as number[]);
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/components/Layout.tsx:194]: Right, I believe the initial state would not be properly displayed because the "names" are not matching the "values" of the select-options.

Can you help create another list `LOG_LEVEL_VALUES_LIST` to be used here?

File: new-log-viewer/src/services/LogFileManager.ts (hunk 2 above)
Comment (junhaoliao) [reply new-log-viewer/src/services/LogFileManager.ts]: The null check was necessary or accessing `result`(null) would cause an error

File: new-log-viewer/src/components/Layout.tsx [hunk 11]
```diff
@@ -187,6 +207,15 @@ const Layout = () => {
         });
     };
 
+    const handleLogLevelSelectChange = (ev: React.ChangeEvent<HTMLSelectElement>) => {
+        const selected = Array.from(ev.target.options)
+            .map((option) => option.selected && Number(option.value))
+            .filter((value) => "number" === typeof value);
+
+        setSelectedLogLevels(selected as number[]);
+        changeLogLevelFilter(selected as unknown as LogLevelFilter);
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/components/Layout.tsx]: Good point. the casts were unnecessary

File: new-log-viewer/src/components/Layout.tsx [hunk 12]
```diff
@@ -277,6 +335,24 @@ const Layout = () => {
                 </button>
 
                 <ConfigForm/>
+
+                <select
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/components/Layout.tsx:339]: We should. Note this `<select/>` component is for debug only and should be replaced by https://mui.com/joy-ui/react-select/ after #59 is completed.

File: new-log-viewer/src/contexts/StateContextProvider.tsx [hunk 13]
```diff
@@ -128,11 +144,16 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
     const {filePath, logEventNum} = useContext(UrlContext);
 
     const [fileName, setFileName] = useState<string>(STATE_DEFAULT.fileName);
+    const [firstLogEventNumPerPage, setFirstLogEventNumPerPage] = useState<number[]>([]);
+    const [lastLogEventNumPerPage, setLastLogEventNumPerPage] = useState<number[]>([]);
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/contexts/StateContextProvider.tsx:148]: Sure. Please also help make the change in `STATE_DEFAULT`.

**Comment by junhaoliao** [comment]
@davemarco I haven't been able to take a careful look into the code yet, but it seems there's one issue whenever I was switching the log level filter from `["TRACE"]` to `["DEBUG"]` - there are exactly 200 `[Renderer -> MainWorker] code=loadPage` requests sent to the service worker, and occasionally the `logEventNum` does not update on editor cursor updates. Could you take a look before we proceed to the review?

Here's an example file I used - 
[example.zip](https://github.com/user-attachments/files/16912955/example.zip)

`pageSize` was set to 10 and I started from http://localhost:3010/?filePath=http://localhost:3010/test/example.jsonl#logEventNum=10

File: new-log-viewer/src/components/Layout.tsx [hunk 14]
```diff
@@ -14,6 +14,9 @@ import APP_THEME from "./theme";
  *
  * @return
  */
+// eslint-disable-next-line no-warning-comments
+// TODO: replace elements in this component once we integrate with a UI framework
+// eslint-disable-next-line max-lines-per-function
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/components/Layout.tsx:19]: Let's remove these. The framework was integrated in #59 and we're using JoyUI elements for the dropdown box.

File: new-log-viewer/src/utils/actions.ts [hunk 15]
```diff
@@ -70,39 +68,43 @@ const EDITOR_ACTIONS : ActionType[] = [
...
  * @param actionName
- * @param logEventNum
- * @param numEvents
+ * @param pageNum
+ * @param numPages
+ * @param loadPage
  */
-const handleAction = (actionName: ACTION_NAME, logEventNum: number, numEvents: number) => {
-    const pageSize = getConfig(CONFIG_KEY.PAGE_SIZE);
+const handleAction = (
```
Comment (junhaoliao) [inline new-log-viewer/src/utils/actions.ts:78]: To avoid creating this extra handler -

Can we change loadPage to have it take in either a `newPage` number or ACTION_NAME? if `typeof newPage` is ACTION_NAME, we calculate the new page number there and proceed to the other logic.

File: new-log-viewer/src/components/MenuBar/PageNumInput.tsx [hunk 16]
```diff
@@ -56,7 +56,7 @@ const PageNumInput = () => {
         }
         inputRef.current.style.width = "0";
         inputRef.current.style.width =
-            `${inputRef.current.scrollWidth + PAGE_NUM_INPUT_FIT_EXTRA_WIDTH}px`;
+         `${inputRef.current.scrollWidth + PAGE_NUM_INPUT_FIT_EXTRA_WIDTH}px`;
```
Comment (junhaoliao) [inline new-log-viewer/src/components/MenuBar/PageNumInput.tsx:59]: Let's revert this change.

File: new-log-viewer/src/components/StatusBar/LogLevelFilter.tsx [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+    const handleChange = (
+        event: React.SyntheticEvent | null,
+        newValue: Array<string> | null
+    ) => {
+    // convert strings to numbers.
+        const selected: LogLevelFilter =
+      newValue && newValue.map((value) => Number(value));
+
+        console.log(newValue);
+
```
Comment (junhaoliao) [inline new-log-viewer/src/components/StatusBar/LogLevelFilter.tsx:46]:
```suggestion
```

File: new-log-viewer/src/components/StatusBar/index.tsx [hunk 18]
```diff
@@ -37,8 +38,12 @@ const StatusBar = () => {
             >
                 Status message
             </Typography>
+            <LogLevelSelect/>
             <Button
                 size={"sm"}
+                sx={{
+                    borderRadius: 0,
```
Comment (junhaoliao) [inline new-log-viewer/src/components/StatusBar/index.tsx:45]: Might be good to create a ".borderless" CSS class and set the `className` of this Button to be `borderless`. What do you think?

**Review by junhaoliao — CHANGES_REQUESTED** [review]
Part 1 / 3

File: new-log-viewer/src/utils/actions.ts (hunk 15 above)
Comment (junhaoliao) [reply new-log-viewer/src/utils/actions.ts:78]: If `loadPage` can handle ACTION_NAME.XXXX_PAGE directly, we can avoid passing the pageNum, the numPages and the loadPage callback all the way here, which might be a bit cleaner. What do you think?

> Is it being used anywhere else?

It's also used by the editor shortcuts if i'm not wrong.

File: new-log-viewer/src/contexts/StateContextProvider.tsx [hunk 19]
```diff
@@ -1,3 +1,6 @@
+// eslint-disable-next-line no-warning-comments
+// TODO: move handlers out from StateContextProvider.
+/* eslint-disable max-lines, max-lines-per-function, max-statements */
```
Comment (junhaoliao) [inline new-log-viewer/src/contexts/StateContextProvider.tsx:3]: Let's remove the `eslint-disable-next-line` for StateContextProvider then.

File: new-log-viewer/src/contexts/StateContextProvider.tsx [hunk 20]
```diff
@@ -133,14 +108,63 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
...
+    const getClosestLogEventNum = useCallback(
+        (beginLineNumToLogEventNum: BeginLineNumToLogEventNumMap) => {
+            let newLogEventNum: Nullable<number> = null;
+            const logEventNumOnPage = Array.from(beginLineNumToLogEventNum.values());
+
+            // On initial file load we don't know logEventNum so just use last value on page
+            if (!logEventNumRef.current) {
+                newLogEventNum = logEventNumOnPage.at(-1) as number;
+
+                return newLogEventNum;
```
Comment (junhaoliao) [inline new-log-viewer/src/contexts/StateContextProvider.tsx:146]:
```suggestion
            return logEventNumOnPage.at(-1) as number;
```

File: new-log-viewer/src/contexts/StateContextProvider.tsx [hunk 21]
```diff
@@ -133,14 +108,63 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
...
+    const pageNumRef = useRef(STATE_DEFAULT.pageNum);
 
     const mainWorkerRef = useRef<null|Worker>(null);
 
+    // Returns the closest logEventNum on the current page to user provided logEventNum. User
+    // provided logEventNum may not be in the filtered array, so here we find the closest value
+    // that actually exists on the page.
+    const getClosestLogEventNum = useCallback(
+        (beginLineNumToLogEventNum: BeginLineNumToLogEventNumMap) => {
+            let newLogEventNum: Nullable<number> = null;
```
Comment (junhaoliao) [inline new-log-viewer/src/contexts/StateContextProvider.tsx:139]:
```suggestion
```

File: new-log-viewer/src/contexts/StateContextProvider.tsx [hunk 22]
```diff
@@ -133,14 +108,63 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
...
+                }
+            }
+
+            if (!newLogEventNum) {
+                // If no nums on page are smaller than user logEventNum, user logEventNum is the
+                //  smallest so we should just return the smallest value on the page.
+                newLogEventNum = logEventNumOnPage[0] as number;
+            }
+
+            return newLogEventNum;
```
Comment (junhaoliao) [inline new-log-viewer/src/contexts/StateContextProvider.tsx:163]:
```suggestion
            // Find first logEventNum smaller or equal to user provided logEventNum.
            const closestLogEventNum = logEventNumOnPage.reverse()
                .find((logEventNum) => logEventNum <= logEventNumRef.current);

            // Return the closest found or fallback to the smallest value on the page.
            return closestLogEventNum ?? logEventNumOnPage[0] as number;
```

File: new-log-viewer/src/contexts/StateContextProvider.tsx [hunk 23]
```diff
@@ -152,10 +176,28 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
...
+                const newPageNum = 1 + firstLogEventNumOnPage.current.findLastIndex(
+                    (value: number) => value <= newLogEventNum,
+                );
+
+                setPageNum(newPageNum);
+                pageNumRef.current = newPageNum;
+
+                updateWindowUrlHashParams({
+                    logEventNum: newLogEventNum,
+                });
```
Comment (junhaoliao) [inline new-log-viewer/src/contexts/StateContextProvider.tsx:193]:
```suggestion
                updateWindowUrlHashParams({logEventNum: newLogEventNum});
```

File: new-log-viewer/src/contexts/StateContextProvider.tsx [hunk 24]
```diff
@@ -275,12 +341,15 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
             value={{
                 beginLineNumToLogEventNum: beginLineNumToLogEventNumRef.current,
                 fileName: fileName,
+                firstLogEventNumOnPage: firstLogEventNumOnPage.current,
+                lastLogEventNumOnPage: lastLogEventNumOnPage.current,
```
Comment (junhaoliao) [inline new-log-viewer/src/contexts/StateContextProvider.tsx:345]: Just so we don't forget this discussion: https://github.com/y-scope/yscope-log-viewer/pull/63/files/7bd410b3514b70e3ff86d0cde2cec5e7fa0f8aea#r1768810124

I'm fine if we want to init the refs with literals. e.g.,
```
    const firstLogEventNumOnPage = useRef<number[]>([]);
    const lastLogEventNumOnPage = useRef<number[]>([]);
```

File: new-log-viewer/src/contexts/StateContextProvider.tsx [hunk 25]
```diff
@@ -166,7 +208,9 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
                 console.error(`Unexpected ev.data: ${JSON.stringify(ev.data)}`);
                 break;
         }
-    }, []);
+    }, [
+        getClosestLogEventNum,
+    ]);
```
Comment (junhaoliao) [inline new-log-viewer/src/contexts/StateContextProvider.tsx:213]:
```suggestion
    }, [getClosestLogEventNum]);
```

File: new-log-viewer/src/contexts/StateContextProvider.tsx [hunk 26]
```diff
@@ -152,10 +176,28 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
...
-                const lastLogEventNum = getLastLogEventNum(args.beginLineNumToLogEventNum);
-                updateLogEventNumInUrl(lastLogEventNum, logEventNumRef.current);
+
+                const newLogEventNum: number = getClosestLogEventNum(
+                    args.beginLineNumToLogEventNum,
+                );
+
+                const newPageNum = 1 + firstLogEventNumOnPage.current.findLastIndex(
+                    (value: number) => value <= newLogEventNum,
+                );
```
Comment (junhaoliao) [inline new-log-viewer/src/contexts/StateContextProvider.tsx:186]: Shall we check the return value of findLastIndex() against -1 ?

File: new-log-viewer/src/contexts/StateContextProvider.tsx [hunk 27]
```diff
@@ -197,55 +251,67 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
...
         numEvents,
         logEventNum,
+        getClosestLogEventNum,
     ]);
 
+    // On `numFilteredEvents` update, set page number to zero if no logs.
+    useEffect(() => {
+        if (0 === numFilteredEvents) {
+            setPageNum(0);
+        }
```
Comment (junhaoliao) [inline new-log-viewer/src/contexts/StateContextProvider.tsx:312]: Shall we use the constants from `STATE_DEFAULT`?

File: new-log-viewer/src/services/LogFileManager.ts [hunk 28]
```diff
@@ -198,6 +213,80 @@ class LogFileManager {
...
+     */
+    changeLogLevelFilter (logLevelFilter: LogLevelFilter) {
+        const result: boolean = this.#decoder.changeLogLevelFilter(logLevelFilter);
+
+        if (false === result) {
+            throw new Error(`Error changing log level filter as feature not yet
+                implemented for this file type on new log viewer`);
+        }
+
+        if (logLevelFilter) {
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/services/LogFileManager.ts:229]: Depending on the type of `logLevelFilter`, `if (logLevelFilter)` can be evaluated to true for a variety of values. 

With that said, now we have type the parameter with TypeScript, the typing has limited the range of possible values that can cause unintended effects.

With those said, those discussions always reminds me of this lol:
> [Explicit is better than implicit.](https://peps.python.org/pep-0020/#:~:text=Explicit%20is%20better%20than%20implicit.)

File: new-log-viewer/src/services/LogFileManager.ts [hunk 29]
```diff
@@ -1,9 +1,12 @@
+/* eslint-disable max-lines */
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/services/LogFileManager.ts:1]: If we're really short of time, let's consider changing the limit instead of completely disabling the rule:
```
/* eslint max-lines: ["error", 400] */
```

File: new-log-viewer/src/services/decoders/JsonlDecoder.ts [hunk 30]
```diff
@@ -24,61 +29,95 @@ import LogbackFormatter from "../formatters/LogbackFormatter";
...
-        // constructor.
+    getFilteredLogEventIndices (): number[] {
+        return this.#filteredLogEventIndices;
+    }
 
-        if (0 > beginIdx || this.#logEvents.length < endIdx) {
-            return null;
+    setLogLevelFilter (logLevelFilter: LogLevelFilter): boolean {
+        this.#filterLogs(logLevelFilter);
+        this.#isFiltered = Boolean(logLevelFilter);
```
Comment (junhaoliao) [inline new-log-viewer/src/services/decoders/JsonlDecoder.ts:70]: Following the discussion at https://github.com/y-scope/yscope-log-viewer/pull/63/files#r1768464251 , I feel this can be more readable:
```suggestion
        this.#isFiltered = null !== logLevelFilter;
```

**Comment by kirkrodrigues** [comment]
Thanks for the additional comments, @junhaoliao. Since this PR ended up being quite large and we are reworking the design, @davemarco is going to submit smaller PRs rather than trying to refactor this one completely.

File: new-log-viewer/src/typings/decoders.ts [hunk 31]
```diff
@@ -13,13 +20,22 @@ interface LogEventCount {
...
     timestampKey: string,
 }
 
-type DecoderOptionsType = JsonlDecoderOptionsType;
+type DecoderOptions = JsonlDecoderOptions;
+
+/**
+ * A log event parsed from a JSON log.
+ */
+interface JsonLogEvent {
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/typings/decoders.ts:34]: Can we make `timestamp` a Dayjs object? That way we will have at least two fields in common. With that said, I agree that those objects are internal to the decoders' own implementations, and we don't have needs for a unified interface yet.

File: new-log-viewer/src/typings/worker.ts [hunk 32]
```diff
@@ -40,11 +43,13 @@ type BeginLineNumToLogEventNumMap = Map<number, number>;
...
     LOAD_FILE = "loadFile",
     LOAD_PAGE = "loadPage",
+    CHANGE_FILTER = "changeFilter",
 }
 
 enum WORKER_RESP_CODE {
-    LOG_FILE_INFO = "fileInfo",
+    LOG_FILE_INFO = "logFileInfo",
     PAGE_DATA = "pageData",
+    VIEW_INFO = "viewInfo",
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/typings/worker.ts:52]: fwiw when I wrote this I had `FILTERED_VIEW_INFO` and `VISIBLE_VIEW_INFO` in mind.

File: new-log-viewer/tsconfig.json [hunk 33]
```diff
@@ -17,11 +17,11 @@
     // "disableReferencedProjectLoad": true,             /* Reduce the number of projects loaded automatically by TypeScript. */
 
     /* Language and Environment */
-    "target": "ES2022",
+    "target": "ES2023",
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/tsconfig.json:20]: I think we need this for some newly implemented Array methods like `findLastIndex`.

**Comment by junhaoliao** [comment]
@kirkrodrigues @davemarco i think we can close this now, right?
